File: src/cpp/ee/ads/internal/GuardedAdView.cpp
#include "ee/ads/internal/GuardedAdView.hpp"

#include <ee/core/ObserverHandle.hpp>
#include <ee/core/Task.hpp>

namespace ee {
namespace ads {
using Self = GuardedAdView;

Self::GuardedAdView(const std::shared_ptr<IAdView>& ad)
    : ad_(ad) {
    loading_ = false;
    loaded_ = false;

    handle_ = std::make_unique<ObserverHandle>();
    handle_->bind(*ad_).addObserver({
        .onLoaded =
            [this] {
                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onLoaded) {
                        observer.onLoaded();
                    }
                });
            },
        .onClicked =
            [this] {
                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onClicked) {
                        observer.onClicked();
                    }
                });
            },
    });
}

Self::~GuardedAdView() = default;

void Self::destroy() {
    ad_->destroy();
    handle_->clear();
}

bool Self::isLoaded() const {
    return loaded_;
}

Task<bool> Self::load() {
    if (loaded_) {
        co_return true;
    }
    if (loading_) {
        // Waiting.
        co_return co_await ad_->load();
    }
    loading_ = true;
    loaded_ = co_await ad_->load();
    loading_ = false;
    co_return loaded_;
}

std::pair<float, float> Self::getAnchor() const {
    return ad_->getAnchor();
}

void Self::setAnchor(float x, float y) {
    ad_->setAnchor(x, y);
}

std::pair<float, float> Self::getPosition() const {
    return ad_->getPosition();
}

void Self::setPosition(float x, float y) {
    ad_->setPosition(x, y);
}

std::pair<float, float> Self::getSize() const {
    return ad_->getSize();
}

void Self::setSize(float width, float height) {
    ad_->setSize(width, height);
}

bool Self::isVisible() const {
    return ad_->isVisible();
}

void Self::setVisible(bool visible) {
    ad_->setVisible(visible);
}
} // namespace ads
} // namespace ee

File: src/unity/Runtime/Ads/Internal/GuardedAdView.cs
using System.Threading.Tasks;

namespace EE.Internal {
    internal class GuardedAdView
        : ObserverManager<IAdViewObserver>, IAdView {
        private readonly IAdView _ad;
        private bool _loading;
        private readonly ObserverHandle _handle;

        public GuardedAdView(IAdView ad) {
            _ad = ad;
            _handle = new ObserverHandle();
            _handle.Bind(ad).AddObserver(new IAdViewObserver {
                OnLoaded = () => DispatchEvent(observer => observer.OnLoaded?.Invoke()),
                OnClicked = () => DispatchEvent(observer => observer.OnClicked?.Invoke())
            });
        }

        public void Destroy() {
            _ad.Destroy();
            _handle.Clear();
        }

        public bool IsLoaded { get; private set; }

        public async Task<bool> Load() {
            if (IsLoaded) {
                return true;
            }
            if (_loading) {
                // Waiting.
                return await _ad.Load();
            }
            _loading = true;
            IsLoaded = await _ad.Load();
            _loading = false;
            return IsLoaded;
        }

        public (float, float) Anchor {
            get => _ad.Anchor;
            set => _ad.Anchor = value;
        }

        public (float, float) Position {
            get => _ad.Position;
            set => _ad.Position = value;
        }

        public (float, float) Size {
            get => _ad.Size;
            set => _ad.Size = value;
        }

        public bool IsVisible {
            get => _ad.IsVisible;
            set => _ad.IsVisible = value;
        }
    }
}